Kernel/Devices/Storage/USB/MassStorageDevice.cpp

```diff
@@ -0,0 +1,134 @@
+#include <Kernel/Devices/Storage/USB/MassStorageDevice.h>
+#include <Kernel/Devices/Storage/USB/SCSIComands.h>
+#include <Kernel/Devices/Storage/USB/USBStorageController.h>
+#include <Kernel/FileSystem/SysFS/Subsystems/Bus/USB/DeviceInformation.h>
+
+namespace Kernel::USB {
+
+ErrorOr<NonnullLockRefPtr<MassStorageDevice>> MassStorageDevice::try_create_from_device(USB::Device const& base_device)
+{
+    auto pipe = TRY(ControlPipe::create(base_device.controller(), 0, base_device.device_descriptor().max_packet_size, base_device.address()));
+    auto device = TRY(adopt_nonnull_lock_ref_or_enomem(new (nothrow) MassStorageDevice(base_device, move(pipe))));
+    TRY(device->m_sysfs_device_info_node.with([&device](auto& node) -> ErrorOr<void> {
+        node = TRY(SysFSUSBDeviceInformation::create(*device));
+        return {};
+    }));
+
+    auto const& configuration = device->configurations().first();
+
+    // Set the configuration to the MassStorage configuration
+    TRY(device->m_default_pipe->submit_control_transfer(
+        USB_REQUEST_RECIPIENT_DEVICE | USB_REQUEST_TYPE_STANDARD | USB_REQUEST_TRANSFER_DIRECTION_HOST_TO_DEVICE,
+        USB_REQUEST_SET_CONFIGURATION, configuration.configuration_id(), 0, 0, nullptr));
+
+    dmesgln("Mass Storage Device with {} interfaces:", configuration.interface_count());
+    for (auto i = 0; i < configuration.interface_count(); ++i) {
+        auto const& interface = configuration.interfaces()[i];
+        auto const& interface_descriptor = interface.descriptor();
+        dmesgln("  {} - id:{}:", i, interface.descriptor().interface_id);
+        dmesgln("    Subclass: {} [{:#02x}]", MassStorage::subclass_string((MassStorage::SubclassCode)interface_descriptor.interface_sub_class_code), interface_descriptor.interface_sub_class_code);
+        dmesgln("    Protocol: {} [{:#02x}]", MassStorage::transport_protocol_string((MassStorage::TransportProtocol)interface_descriptor.interface_protocol), interface_descriptor.interface_protocol);
+
+        if (interface_descriptor.interface_protocol == to_underlying(MassStorage::TransportProtocol::BBB)
+            && interface_descriptor.interface_sub_class_code == to_underlying(MassStorage::SubclassCode::SCSI_transparent)) {
+            // We have a Bulk storage we know how to talk to
+            u8 max_luns;
+            TRY(device->m_default_pipe->submit_control_transfer(
+                USB_REQUEST_TYPE_CLASS | USB_REQUEST_RECIPIENT_INTERFACE | USB_REQUEST_TRANSFER_DIRECTION_DEVICE_TO_HOST,
+                (u8)MassStorage::RequestCodes::GetMaxLun, 0, interface.descriptor().interface_id, 1, &max_luns));
+            // FIXME: Devices that do not support multiple LUNs may STALL this command
+            // FIXME: Support multiple LUNs
+            if (max_luns != 0)
+                dmesgln("WARNING: USB Mass Storage Device support multiple LUNs ({}) only targetting first LUN", max_luns);
+
+            u8 in_pipe_address = -1;
```

Comment:
This feels weird, why not just use `0xff`?

Reply:
`-1` looks more like an invalid state than `0xff`, imo

Reply:
But it also doesn't fit in an `u8` (it gets wrapped around) and falsely suggests `in_pipe_address` actually stores that value. Why not be clear about the actual value it contains?

Reply:
Still disagree but changed that to `0xff`